File: leet/0116_populating_next_pointer/0116_populating_next_pointer.py
# LeetCode 116. Populating Next Right Pointers in Each Node
# https://leetcode.com/problems/populating-next-right-pointers-in-each-node/
# O(n) time, O(1) space
class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if root and root.left and root.right:
            root.left.next = root.right
            if root.next:
                root.right.next = root.next.left
            self.connect(root.left)
            self.connect(root.right)
        return root
        

File: leet/0116_populating_next_pointer/0116_populating_next_pointer.ts
function connect(root: Node | null): Node | null {
    const connectViaParent = (node: Node, parent: Node) => {
        const nextParent = node;
        while (node) {
            node.next = parent.right;
            if (parent.next)
                node.next.next = parent.next.left;
            node = node.next.next;
            parent = parent.next;
        }
        if (nextParent)
            connectViaParent(nextParent.left, nextParent);
    }
    if (root)
        connectViaParent(root.left, root);
    return root;
};
